ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,381 @@
+#include <string.h>
+#include "Locks.hpp"
+#include "VnodeCache.hpp"
+#include "Memory.hpp"
+#include "KextLog.hpp"
+
+#ifdef KEXT_UNIT_TESTING
+#include "VnodeCacheTestable.hpp"
+#endif
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked();
+static inline void UpgradeToExclusiveLock(RWLock& lock);
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode);
+
+KEXT_STATIC bool TryFindVnodeIndex_SharedLocked(
+    vnode_t _Nonnull vnode,
+    uintptr_t startingIndex,
+    uintptr_t stoppingIndex,
+    /* out parameters */
+    uintptr_t& cacheIndex);
+
+KEXT_STATIC void UpdateCacheEntryToLatest_ExclusiveLocked(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    uintptr_t index,
+    vnode_t _Nonnull vnode,
+    const FsidInode& vnodeFsidInode,
+    uint32_t vnodeVid);
+
+KEXT_STATIC bool FindAndUpdateEntryToLatest_ExclusiveLocked(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    const FsidInode& vnodeFsidInode,
+    uintptr_t startingIndex,
+    uintptr_t stoppingIndex,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out paramaeters */
+    VirtualizationRootHandle& rootHandle);
+
+static void ClearCacheAndInsertEntry_ExclusiveLocked(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    const FsidInode& vnodeFsidInode,
+    uintptr_t startingIndex,
+    uint32_t vnodeVid,
+    /* out paramaeters */
+    VirtualizationRootHandle& rootHandle);
+
+struct VnodeCacheEntry
+{
+    vnode_t vnode;
+    uint32_t vid;   // vnode generation number
+    VirtualizationRootHandle virtualizationRoot;
+};
+
+KEXT_STATIC uint32_t s_entriesCapacity;
+KEXT_STATIC VnodeCacheEntry* s_entries;
+static RWLock s_entriesLock;
+static const uint32_t MinEntriesCapacity = 0x040000; //  4 MB (assuming 16 bytes per VnodeCacheEntry)
+static const uint32_t MaxEntriesCapacity = 0x400000; // 64 MB (assuming 16 bytes per VnodeCacheEntry)
+
+kern_return_t VnodeCache_Init()
+{
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+    
+    s_entriesLock = RWLock_Alloc();
+    if (!RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+
+    s_entriesCapacity = MAX(MinEntriesCapacity, desiredvnodes * 2);
+    s_entriesCapacity = MIN(MaxEntriesCapacity, s_entriesCapacity);
+    
+    s_entries = Memory_AllocArray<VnodeCacheEntry>(s_entriesCapacity);
+    if (nullptr == s_entries)
+    {
+        s_entriesCapacity = 0;
+        return KERN_RESOURCE_SHORTAGE;
+    }
+    
+    VnodeCache_InvalidateCache(nullptr);
```

Comment:
`Memory_AllocArray` returns uninitialised memory, and invalidating the cache happens to be a memset. I'm not 100% sold on how the vnode cache code mixes business logic with low-level memory management and hash table operations, but I guess that debate is why #500 is still open…

Reply:
> Why do we have to invalidate cache to init?
> invalidating the cache happens to be a memset

I agree this looks weird and it requires the caller of `VnodeCache_InvalidateCache` to make assumptions about the implementation of `VnodeCache_InvalidateCache`.

I will make this more explicit so that if we choose to change the implementation of `VnodeCache_InvalidateCache` in the future it will not break `VnodeCache_Init`.